Improve tracing architecture documentation (#5958)
kyma-project/kyma: docs/tracing/02-01-tracing.md

title: Architecture
---

The Jaeger-based tracing component provides the necessary functionality to collect and query traces. Both operations may occur at the same time. This way you  you can inspect specific traces using the Jaeger UI, while Jaeger takes care of proper trace collection and storage in parallel. See the diagram for details: 

![Tracing architecture](./assets/tracing-architecture.svg)

## Collect traces

The process of collecting traces by Jaeger looks as follows:
 
1. The application receives a request, either from an internal or external source.

> 2. If the application has Istio injection enabled, [Istio proxy](https://github.com/istio/proxy) propagates the correct [HTTP headers](/components/tracing#details-propagate-http-headers) of the requests to the Jaeger Deployment. Istio proxy accesses the Deployment using the [Zipkin](https://zipkin.io/) service which exposes a Jaeger port compatible with the Zipkin protocol.  

Reviewer: ```suggestion
2. If the application has Istio injection enabled, [Istio proxy](https://github.com/istio/proxy) propagates the correct [HTTP headers](/components/tracing#details-propagate-http-headers) of the requests to the Jaeger Deployment. Istio proxy calls Jaeger using the [Zipkin](https://zipkin.io/) service which exposes a port compatible with the Zipkin protocol.  
```

3. Jaeger processes the data. Specifically, the  Jaeger Agent component receives the spans, batches them, and forwards to the Jaeger Collector service. 

> 4. The BadgerDB database stores the data and persists it using the PersistentVolume resource.

Reviewer: ```suggestion
4. The BadgerDB database stores the data and persists it using a PersistentVolume resource.
```

title: Architecture
---

In Kyma, the tracing component based on Jaeger provides the necessary functionality to collect and query traces. Both operations may occur at the same time, meaning that you can inspect specific traces using the Jeager UI, while at the same time Jaeger takes care of proper trace collection and storage. See the diagram for details: 

> ![Tracing architecture](./assets/tracing-architecture.svg)

Reviewer 2: I noticed that Jaeger is capitalized and sometimes not - not sure if that is how it should be?
![image](https://user-images.githubusercontent.com/40655785/66658012-ea088600-ec40-11e9-859e-7c4ab61c076c.png)

## Collect traces

The process of collecting traces by Jaeger looks as follows:
 
1. The application receives a request, either from an internal or external source.

> 2. If the application has Istio injection enabled, [Istio proxy](https://github.com/istio/proxy) makes sure to propagate the correct [HTTP headers](/components/tracing#details-propagate-http-headers) of the requests further to the Jaeger Deployment. Istio proxy accesses the Deployment using the [Zipkin](https://zipkin.io/) service which exposes a Jaeger port compatible with the Zipkin protocol.  

Reviewer 2: Does Istio proxy propagate the correct headers itself? (then we could get rid of `make sure`)

title: Architecture
---

In Kyma, tracin component provides the necessary functionality to collect and query traces. Both operations may occur at the same time, meaning that you can retrieve specific traces using the Jeager UI, while at the same time Jaeger and [Zipkin](https://zipkin.io/) take care of proper trace collection ans storage. See the diagram for details: 

![Tracing architecture](./assets/tracing-architecture.svg)

## Collect traces

The process of collecting traces by Jaeger looks as follows:
 
1. The application receives trace data.

> 2. Istio Proxy configured for the application makes sure to propagate the correct [HTTP headers](/components/tracing#details-propagate-http-headers) of the requests further to the `zipkin` service, responsible for looking up and collecting traces from various distributed systems.

Reviewer: ```suggestion
2. If the application has istio-injection enabled, the Istio Proxy for the application makes sure to propagate the correct [HTTP headers](/components/tracing#details-propagate-http-headers) of the requests further to the Jaeger deployment accessing it via the `zipkin` service, exposing a Jaeger port compatible to the zipkin protocol.
```

> 3. Zipkin receives the data and passes it further through [Istio Proxy](https://github.com/istio/proxy) which provides further authorization mechanisms to ensure proper communication. 

Reviewer: ```suggestion

```

> 4. Jaeger processes the data. Specifically, the  `jaeger-agent` service receives the spans, batches them and forwards to collectors. 

Reviewer: ```suggestion
4. Jaeger processes the data. Specifically, the  `agent`-component receives the spans, batches them and forwards to the `collectors?-component. 
```

> 5. The data is then saved in the database.

Reviewer: ```suggestion
5. The data is then saved in a database (BadgerDB) persisted on a persistent volume.
```

## Query traces

The process of querying traces from Jaeger looks as follows:

1. A Kyma user accesses the Jaeger UI to [look for specific traces](/components/tracing#details-search-for-traces).

> 2. Jaeger UI passes the request to the `jaeger-query` service. The requests goes through the [Istio Ingress Gateway](https://kyma-project.io/docs/components/application-connector/#architecture-application-connector-components-istio-ingress-gateway) which receives and processes the incoming connections between the service and the external source.

Reviewer: ```suggestion
2. Jaeger UI passes the request to the `jaeger-query` service. The requests goes through the [Istio Ingress Gateway](https://kyma-project.io/docs/components/application-connector/#architecture-application-connector-components-istio-ingress-gateway) which forwards the incoming connections to the service.
```

> 3. The `jaeger-query` service passes the request to the [Keycloak Gatekeeper](https://github.com/keycloak/keycloak-gatekeeper) for verification. The Gatekeeper calls [Dex](https://github.com/dexidp/dex) to authenticate the user and the request, and grants further access if the authentication is successful. 

Reviewer: ```suggestion
3. The `jaeger-query` service passes the request to the [Keycloak Gatekeeper](https://github.com/keycloak/keycloak-gatekeeper) for request authorization. The Gatekeeper calls [Dex](https://github.com/dexidp/dex) to authenticate the user and the request, and grants further access if the authentication is successful. 
```

> 4. The request goes through [Istio Proxy](https://github.com/istio/proxy) which provides further authorization mechanisms to ensure proper communication. 

Reviewer: ```suggestion

```
